File: src/vm/bytecode.py
"""
ClaudeLang Bytecode Definitions

Stack-based bytecode for efficient execution.
"""

from enum import IntEnum, auto
from dataclasses import dataclass
from typing import Any, List, Optional


class Opcode(IntEnum):
    """Bytecode opcodes for ClaudeLang VM"""
    
    # Stack manipulation
    PUSH = auto()       # Push constant onto stack
    POP = auto()        # Pop and discard top
    DUP = auto()        # Duplicate top
    SWAP = auto()       # Swap top two elements
    
    # Arithmetic
    ADD = auto()        # Pop 2, push sum
    SUB = auto()        # Pop 2, push difference
    MUL = auto()        # Pop 2, push product
    DIV = auto()        # Pop 2, push quotient
    MOD = auto()        # Pop 2, push remainder
    NEG = auto()        # Pop 1, push negation
    
    # Comparison
    EQ = auto()         # Pop 2, push equality
    NE = auto()         # Pop 2, push inequality
    LT = auto()         # Pop 2, push less than
    LE = auto()         # Pop 2, push less or equal
    GT = auto()         # Pop 2, push greater than
    GE = auto()         # Pop 2, push greater or equal
    
    # Boolean
    AND = auto()        # Pop 2, push logical and
    OR = auto()         # Pop 2, push logical or
    NOT = auto()        # Pop 1, push logical not
    
    # Control flow
    JUMP = auto()       # Unconditional jump
    JUMP_IF = auto()    # Pop 1, jump if true
    JUMP_IF_NOT = auto()# Pop 1, jump if false
    CALL = auto()       # Function call
    RETURN = auto()     # Return from function
    
    # Variables
    LOAD = auto()       # Load variable
    STORE = auto()      # Store variable
    LOAD_GLOBAL = auto()# Load global
    STORE_GLOBAL = auto()# Store global
    
    # Functions
    MAKE_FUNC = auto()  # Create closure
    MAKE_ENV = auto()   # Create new environment
    POP_ENV = auto()    # Restore previous environment
    
    # Lists
    MAKE_LIST = auto()  # Create list from N stack items
    LIST_HEAD = auto()  # Get first element
    LIST_TAIL = auto()  # Get rest of list
    LIST_CONS = auto()  # Cons element to list
    LIST_LEN = auto()   # Get list length
    LIST_EMPTY = auto() # Check if list empty
    
    # Strings
    STR_LEN = auto()    # String length
    STR_CONCAT = auto() # Concatenate strings
    STR_UPPER = auto()  # Uppercase string
    STR_LOWER = auto()  # Lowercase string
    
    # Effects
    EFFECT = auto()     # Trigger effect
    
    # Special
    HALT = auto()       # Stop execution
    NOP = auto()        # No operation


@dataclass
class Instruction:
    """Single bytecode instruction"""
    opcode: Opcode
    arg: Optional[Any] = None
    
    def __repr__(self):
        if self.arg is not None:
            return f"{self.opcode.name} {self.arg}"
        return self.opcode.name


@dataclass
class BytecodeChunk:
    """A chunk of bytecode with constants"""
    instructions: List[Instruction]
    constants: List[Any]
    
    def add_instruction(self, opcode: Opcode, arg: Optional[Any] = None):
        """Add an instruction"""
        self.instructions.append(Instruction(opcode, arg))
    
    def add_constant(self, value: Any) -> int:
        """Add a constant and return its index"""
        # Reuse existing constants
        if value in self.constants:
            return self.constants.index(value)
        
        self.constants.append(value)
        return len(self.constants) - 1
    
    def disassemble(self) -> str:
        """Disassemble bytecode to readable format"""
        lines = ["=== Bytecode ==="]
        
        # Constants
        if self.constants:
            lines.append("\nConstants:")
            for i, const in enumerate(self.constants):
                lines.append(f"  [{i}] {repr(const)}")
        
        # Instructions
        lines.append("\nInstructions:")
        for i, instr in enumerate(self.instructions):
            lines.append(f"  {i:04d}: {instr}")
        
        return "\n".join(lines)


class BytecodeBuilder:
    """Helper for building bytecode"""
    
    def __init__(self):
        self.chunk = BytecodeChunk([], [])
    
    def emit(self, opcode: Opcode, arg: Optional[Any] = None):
        """Emit an instruction"""
        self.chunk.add_instruction(opcode, arg)
    
    def emit_constant(self, value: Any):
        """Emit a constant push"""
        idx = self.chunk.add_constant(value)
        self.emit(Opcode.PUSH, idx)
    
    def emit_jump(self, opcode: Opcode) -> int:
        """Emit a jump instruction and return position for backpatching"""
        pos = len(self.chunk.instructions)
        self.emit(opcode, 0)  # Placeholder
        return pos
    
    def patch_jump(self, pos: int, target: int):
        """Patch a jump instruction with target"""
        self.chunk.instructions[pos].arg = target
    
    def current_position(self) -> int:
        """Get current instruction position"""
        return len(self.chunk.instructions)
    
    def get_chunk(self) -> BytecodeChunk:
        """Get the completed chunk"""
        return self.chunk

File: rust/fluentai-vm/src/bytecode.rs
//! Bytecode representation for FluentAi VM

use fluentai_core::value::Value;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Opcode {
    // Stack manipulation
    Push,
    Pop,
    Dup,
    Swap,

    // Arithmetic
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Neg,

    // Specialized unboxed arithmetic (faster paths)
    AddInt,   // Add two integers without boxing
    SubInt,   // Subtract two integers
    MulInt,   // Multiply two integers
    DivInt,   // Divide two integers
    AddFloat, // Add two floats
    SubFloat, // Subtract two floats
    MulFloat, // Multiply two floats
    DivFloat, // Divide two floats

    // Comparison
    Eq,
    Ne,
    Lt,
    Le,
    Gt,
    Ge,

    // Type-specialized comparison
    LtInt,
    LeInt,
    GtInt,
    GeInt,

    // Boolean
    And,
    Or,
    Not,

    // Control flow
    Jump,
    JumpIf,
    JumpIfNot,
    Call,
    Return,

    // Variables
    Load,
    Store,
    LoadGlobal,
    StoreGlobal,

    // Fast local variable access
    LoadLocal0,
    LoadLocal1,
    LoadLocal2,
    LoadLocal3,
    StoreLocal0,
    StoreLocal1,
    StoreLocal2,
    StoreLocal3,

    // Functions
    MakeFunc,
    MakeClosure,  // Like MakeFunc but captures N values from stack
    LoadCaptured, // Load value from captured environment
    MakeEnv,
    PopEnv,

    // Let binding cleanup
    // Removes N values from stack but preserves the top value
    PopN,

    // Lists
    MakeList,
    ListHead,
    ListTail,
    ListCons,
    ListLen,
    ListEmpty,

    // Strings
    StrLen,
    StrConcat,
    StrUpper,
    StrLower,

    // Specialized constants
    PushInt0,
    PushInt1,
    PushInt2,
    PushIntSmall, // for small integers that fit in arg
    PushTrue,
    PushFalse,
    PushNil,
    PushConst, // Push constant from constant pool

    // Effects
    Effect,
    EffectAsync,
    Await,
    Spawn,
    Channel,
    Send,
    Receive,

    // Effect handlers
    MakeHandler,      // Create handler table from stack values
    InstallHandler,   // Install handler for dynamic scope
    UninstallHandler, // Uninstall handler, preserving result

    // Mutable cells
    MakeCell, // Create a cell with initial value
    CellGet,  // Get value from cell
    CellSet,  // Set value in cell

    // Tagged values (for constructor patterns)
    MakeTagged,     // Create tagged value: tag(string const), arity -> Tagged
    GetTag,         // Get tag from tagged value
    GetTaggedField, // Get field N from tagged value
    IsTagged,       // Check if value is tagged with specific tag

    // Module operations
    LoadModule,    // Load module by name (string const)
    ImportBinding, // Import specific binding from module
    ImportAll,     // Import all exports from module
    LoadQualified, // Load qualified variable (module.name)
    BeginModule,   // Mark beginning of module scope
    EndModule,     // Mark end of module scope
    ExportBinding, // Export a binding from current module

    // GC operations
    GcAlloc,   // Allocate value with GC
    GcDeref,   // Dereference GC handle
    GcSet,     // Set value in GC handle
    GcCollect, // Manually trigger collection

    // Tail call optimization
    TailCall,    // Tail call with frame reuse
    TailReturn,  // Return from tail-recursive function
    LoopStart,   // Mark start of tail-recursive loop
    LoopEnd,     // Mark end of tail-recursive loop
    UpdateLocal, // Update local variable (for loop parameter updates)

    // Special
    Halt,
    Nop,
}

#[derive(Debug, Clone)]
pub struct Instruction {
    pub opcode: Opcode,
    pub arg: u32,
}

impl Instruction {
    pub fn new(opcode: Opcode) -> Self {
        Self { opcode, arg: 0 }
    }

    pub fn with_arg(opcode: Opcode, arg: u32) -> Self {
        Self { opcode, arg }
    }
}

// Value type is now imported from fluentai_core::value::Value
// The core Value type includes all the variants needed by the VM

// Display implementation is provided by fluentai_core::value::Value

#[derive(Debug, Clone)]
pub struct BytecodeChunk {
    pub instructions: Vec<Instruction>,
    pub constants: Vec<Value>,
    pub name: Option<String>,
}

impl BytecodeChunk {
    pub fn new(name: Option<String>) -> Self {
        Self {
            instructions: Vec::new(),
            constants: Vec::new(),
            name,
        }
    }

    pub fn add_instruction(&mut self, instruction: Instruction) -> usize {
        self.instructions.push(instruction);
        self.instructions.len() - 1
    }

    pub fn add_constant(&mut self, value: Value) -> u32 {
        self.constants.push(value);
        (self.constants.len() - 1) as u32
    }

    pub fn patch_jump(&mut self, offset: usize, target: usize) {
        assert!(offset < self.instructions.len(), "Invalid jump offset");
        assert!(target <= u32::MAX as usize, "Jump target overflow");
        self.instructions[offset].arg = target as u32;
    }
}

#[derive(Debug, Clone)]
pub struct Bytecode {
    pub chunks: Vec<BytecodeChunk>,
    pub main_chunk: usize,
}

impl Bytecode {
    pub fn new() -> Self {
        Self {
            chunks: Vec::new(),
            main_chunk: 0,
        }
    }

    pub fn add_chunk(&mut self, chunk: BytecodeChunk) -> usize {
        self.chunks.push(chunk);
        self.chunks.len() - 1
    }
}
